### ISSUE TITLE
Early crash when xvi is linked to termcap library.

### ISSUE THREAD

**Issue by jvvv** [martinwguy/xvi#63]
When testing, I usually link against ncurses or netbsd-curses (actually, libterminfo from that package). Recently, I tried linking against GNU termcap library. The resultant binary crashes on startup while reading capabilities. Termcap only works with what ever buffer we give it when we call tgetent, which we've sized at 1024 bytes in tty_open() from src/tcap_scr.c:798. So when the termcap entry is longer that 1024 bytes, we get a crash while trying to retrieve capabilities. This isn't a problem when linking against curses which uses compatibility functions that ignore the buffer space we give it and allocate their own.

Quick fix is to increase the size of tcbuf. I added a debug printf just after the check for tgetent failure that prints the strlen of tcbuf after the call. For my usual TERM (xterm-256color) the length is 3208. I'm tempted to replace the array with dynamic calls:
```
char *tcbuf;
tcbuf = alloc(4096);
...
switch (tgetent(tcbuf, termtype)) {
...
}
tcbuf = re_alloc(tcbuf, strlen(tcbuf)+1);
```
The quickfix is easier and fewer loc, but using the dynamic is more memory effecient.

**Comment by martinwguy** [comment]
The quick fix is fine, as it reallocs to the amount it actually needs
to use, freeing the surplus. If we're that short of memory at startup,
the edito 's unlikely to get much further anyway.

**Comment by jvvv** [comment]
Some how I've managed to say only half of what I meant to say, and in the balance miss a few other points to ponder.

What I meant by quick fix was simply:
```
char tcbuf[4096];
```

The reason I didn't concider the alloc/re_alloc scheme easy is that now we have bit of malloc'd space tied to a system specific routine. If ever we decide to try to work on reducing/eliminating memory leaks, here is a potential one.

Another point to concider is the termcap/terminfo differences. If the underlying termcap routines are actually terminfo's termcap compatibility functions, then we are compounding the memory usage with either fix. I intend to see if there are any reliable macros we can use to tell whether we have true termcap or curses/terminfo compatibility termcap. Then we could conditionally compile the appropriate tgetent call to match our terminal capablility library. So far, though, I'm not having any luck on find such a macro.

**Comment by jvvv** [comment]
What I didn't point out is that the tgetent call in the terminfo library (or curses) ignores the buffer argument and uses/allocates it's own, so any space we set aside or allocate is purely wasted in this case.

**Comment by martinwguy** [comment]
Yes, sorry, I wasn't very clear. I meant yes to the alloc(4096) - re_alloc() code you quote above, The current static buffer, whatever size it has, if termcap is wasteful and if terminfo is a complete waste. Well spotted!

I remember in the 1980s measuring the max size of termcap entries and 1024 was above the max. The web docs on GNU termcap have an example using 2048, but you say you've found a terminal tha uses 3208 (do terminals really do that much more these days?!) so 4096 should keep us safe for a few more years. Like I say, if it can't allocate the extra 900 bytes even before it's read in their text file, then it probably won't be usable anyway.

If terminfo's termcap shim doesn't need the buffer at all and you can reliably detect that, so much the better. The not-needing-the-buffer behaviour is documented so we can use it, but I haven't yet seen how to find out whether curs_termcap is the implementation in use.

For GNU termcap instead, I see:
"If you are using the GNU version of termcap, you can alternatively ask tgetent to allocate enough space. Pass a null pointer for buffer, and tgetent itself allocates the storage using malloc. There is no way to get the address that was allocated, and you shouldn't try to free the storage."
though I also don't see how you find out whether you are using GNU termcap!

Is it guaranteed that the stuff tgetent writes into the buffer is null-terminated and only significant up to and including the first nul? Unfortunately I don't see any docs about max buffer size, nothing to tell you how much is used at runtime, and no way to protect against overflow, which only leaves giving it an immense buffer and hoping for the best.
Worst comes to the worst, if not GNU and not curs_termcap, I suppose one could find out how much buffer it's using by writing a canary value (like 0xAA, or maybe NUL which is free) into it before calling tgetent, then realloc to however much it's overwritten...

**Comment by martinwguy** [comment]
Incidentally, GNU termcap's README says:

"This package contains termcap.src, the latest official termcap data
file.  By default, it is not installed.  The current version contains
some entries that are more than 1023 bytes long, which is the largest
value that is safe to use with the many historical applications that
only allocate a 1024 byte termcap buffer (telnet, for example).  If
you make sure that all of your programs allocate buffers of at least
2500 bytes, or let the termcap library do it by passing a NULL
pointer, then it is safe to install the new termcap file, as described
below."

Which TERM entry used over 3K?

However, the tcbuf[1024, 2500, 4096,whatever] buffer is allocated on the stack in ttyopen(), xvi copies the strings it needs into tty_open's static buffer "strings[512]" and tcbuf is freed from the stack when tty_open() returns. On this system (xterm) it uses 263 of those 512 bytes, so we can save 1/4 K of RAM or more by resizing it.

**Comment by martinwguy** [comment]
...assuming that realloc() doesn't move the buffer, of course, as dozens of character pointers are set to point into it. It's unlikely that realloc() would move the buffer when making it smaller, but it *is* allowed to do so if it wishes, which would make xvi die horribly.

To be safe, I'd just up the tcbuf size to whatever seems best.  Having GNU termcap allocate its own buffer is worse because it means that memory would not be freed, whereas it is if it uses our tcbuf; for the terminfo shims we can't do anything about that but it's harmless.

**Comment by jvvv** [comment]
I agree. That seems the safest way to me also. I will continue to look for a way to determine which implementation we are building against. But I see no reason to leave the possible crash waiting for it.

**Comment by martinwguy** [comment]
Fixed by 570670a
